### ISSUE TITLE
Unit Of Work & Repository Pattern

### ISSUE THREAD

**Issue by Speedplay** [dotnet/docs#15226]
How come the those pattern are so differently used in this example?
In my understanding, the Unit Of Work should contains the repositories.
In this example the Repository is using the Unit Of Work.
I'm a bit confused.

https://docs.microsoft.com/en-us/aspnet/mvc/overview/older-versions/getting-started-with-ef-5-using-mvc-4/implementing-the-repository-and-unit-of-work-patterns-in-an-asp-net-mvc-application

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9c1c070c-b2ff-f29d-21df-bcc7bf9dd39d
* Version Independent ID: 975107e9-0fc2-b867-d917-595412cbe422
* Content: [Implementing the infrastructure persistence layer with Entity Framework Core](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/microservice-ddd-cqrs-patterns/infrastructure-persistence-layer-implemenation-entity-framework-core#feedback)
* Content Source: [docs/architecture/microservices/microservice-ddd-cqrs-patterns/infrastructure-persistence-layer-implemenation-entity-framework-core.md](https://github.com/dotnet/docs/blob/master/docs/architecture/microservices/microservice-ddd-cqrs-patterns/infrastructure-persistence-layer-implemenation-entity-framework-core.md)
* Product: **dotnet-architecture**
* Technology: **microservices**
* GitHub Login: @nishanil
* Microsoft Alias: **nanil**

**Comment by mvelosop** [comment]
Hi @Speedplay, thanks for your feedback!

Actually, _UnitOfWork_ and _Repository_ have different, although related, responsibilities and any of them can exist without the other. However, usually, a _Repository_ **uses** an _UnitOfWork_.

You can think of _UnitOfWork_ as the equivalent of a transaction in the sense that it buffers updates to the the database, to send them in a single batch when saving. _UnitOfWork_ is mostly about change tracking.

You can see a _Repository_ as a collection-based interface to access domain objects so client classes don't have to deal with the persistence details. If you're using _UnitOfWork_ it makes sense for the _Repository_ to get the objects (use) from the _UnitOfWork_. a _Repository_ is mostly about making it easier for client classes to get domain objects.

However, You can have a repository accessing a DB directly, for example using Dapper and sending all changes to the DB as they come through, relying on a DB transaction to rollback in case changes are not to be saved. So a _Repository_ **doesn't always need** to use a _UnitOfWork_ to do its job.

In the referenced article, I see the repositories inside the "UnitOfWork" just as a way to organize the repositories in a class for convenience, to make it evident that all of them rely on the same "context", but that "UnitOfWork" is not really doing any change tracking (so it's not really a _UnitOfWork_).

Actually it looks like it's **SchoolContext** the class that's really taking care of change tracking.

And you can also check that, in that sample, the repositories are **using** SchoolContext, the seemingly real _UnitOfWork_.

So, to summarize:

_Repositories_ usually depend on a _UnitOfWork_ because they use it, not because they have to be part of one, so a _UnitOfWork_ doesn't have to own them (have them in a collection).

You can also check [UnitOfWork](https://www.martinfowler.com/eaaCatalog/unitOfWork.html) and [Repository](https://martinfowler.com/eaaCatalog/repository.html) definitions in the [Patterns of Enterprise Application Architecture book](https://www.martinfowler.com/books/eaa.html) for more details.

Hope this helps.

**Comment by mvelosop** [comment]
Closing this issue now but feel free to comment, will reopen if needed.
